### DISCUSSION TITLE
Race condition in the "Solar heater control"

### DISCUSSION THREAD

**Discussion by steelman** [arendst/Tasmota#20806] [Q&A] [answered]
Hi,

I am implementing something like the [Solar heater control](https://tasmota.github.io/docs/Rules/#solar-heater-control) only without mem3 based threshold.

My rules are

```
Rule1
  ON system#boot DO Backlog Var1 50; Var2 50 ENDON
  ON DS18B20-1#temperature DO event t1=%value% ENDON
  ON DS18B20-2#temperature DO event t2=%value% ENDON
  ON event#t1 DO Backlog Var1 %value% ; ADD1 3 ENDON 
  ON event#t1 DO Backlog Var2 %value% ; ADD2 1 ENDON 
  ON event#t2>%Var1% DO POWER ON ENDON
  ON event#t2<%Var2% DO POWER OFF ENDON
```

The actuator reacts only to the temperature difference. Weird stuff happens, because when the temperature on DS18B20-2 is dropping, the power flickers sometimes, which in theory must not happen when `Var1` and `Var2` are 2°C apart and `t1` and `t2` almost equal. Even in the log the condition is quite obvious. If Tasmota was (is it?) multithreaded I'd assume there is a race between `ADD1`, `ADD2` and `event#t2` checks, but otherwise I can't see a reason for both condition to be met almost at once. Any other ideas?
 
EDIT: Changing to `Tele-DS18B20-1#temperature` and `Tele-DS18B20-2#temperature` with the TelePeriod of 30 seconds fixes the problem.

```
19:22:04.680 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:04.694 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.8"
19:22:04.720 RSL: RESULT = {"Var1":"24.3"}
19:22:04.969 RSL: RESULT = {"Add1":"27.300"}
19:22:05.229 RSL: RESULT = {"Var2":"24.3"}
19:22:05.291 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:05.305 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.8"
19:22:05.329 RSL: RESULT = {"Var1":"24.3"}
19:22:05.579 RSL: RESULT = {"Event":"Done"}
19:22:05.605 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:05.619 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:05.631 RSL: RESULT = {"Event":"Done"}
19:22:05.657 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:05.672 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:05.687 RSL: RESULT = {"Event":"Done"}
19:22:05.713 RUL: EVENT#T2>%VAR1% performs "POWER ON"
19:22:05.718 RSL: RESULT = {"POWER":"ON"}
19:22:05.720 RSL: POWER = ON
19:22:05.899 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:05.914 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.8"
19:22:05.938 RSL: RESULT = {"Var1":"24.3"}
19:22:06.188 RSL: RESULT = {"Add1":"27.300"}
19:22:06.438 RSL: RESULT = {"Var2":"24.3"}
19:22:06.479 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:06.495 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.8"
19:22:06.568 RSL: RESULT = {"Event":"Done"}
19:22:06.596 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:06.611 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:06.626 RSL: RESULT = {"Event":"Done"}
19:22:06.832 RSL: RESULT = {"Var1":"24.3"}
19:22:07.081 RSL: RESULT = {"Add1":"27.300"}
19:22:07.117 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:07.132 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:07.145 RSL: RESULT = {"Var2":"24.3"}
19:22:07.396 RSL: RESULT = {"Event":"Done"}
19:22:07.422 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:07.436 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:07.448 RSL: RESULT = {"Event":"Done"}
19:22:07.698 RSL: RESULT = {"Event":"Done"}
19:22:07.725 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:07.739 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:07.761 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:07.775 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:07.790 RSL: RESULT = {"Event":"Done"}
19:22:08.040 RSL: RESULT = {"Var1":"24.3"}
19:22:08.290 RSL: RESULT = {"Add1":"27.300"}
19:22:08.326 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:08.340 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:08.359 RSL: RESULT = {"Var2":"24.3"}
19:22:08.609 RSL: RESULT = {"Event":"Done"}
19:22:08.640 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:08.655 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:08.693 RSL: RESULT = {"Event":"Done"}
19:22:08.895 RSL: RESULT = {"Var1":"24.3"}
19:22:08.934 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:08.950 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:08.965 RSL: RESULT = {"Add1":"27.300"}
19:22:09.206 RSL: RESULT = {"Var2":"24.3"}
19:22:09.433 RSL: RESULT = {"Add2":"25.300"}
19:22:09.494 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:09.508 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:09.533 RSL: RESULT = {"Var1":"24.3"}
19:22:09.782 RSL: RESULT = {"Event":"Done"}
19:22:09.808 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:09.822 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:09.836 RSL: RESULT = {"Event":"Done"}
19:22:09.862 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:09.876 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:09.888 RSL: RESULT = {"Event":"Done"}
19:22:09.914 RUL: EVENT#T2>%VAR1% performs "POWER ON"
19:22:09.919 RSL: RESULT = {"POWER":"ON"}
19:22:09.921 RSL: POWER = ON
19:22:09.933 RUL: EVENT#T2<%VAR2% performs "POWER OFF"
19:22:09.938 RSL: RESULT = {"POWER":"OFF"}
19:22:09.940 RSL: POWER = OFF
19:22:10.108 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:10.123 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:10.148 RSL: RESULT = {"Var1":"24.3"}
19:22:10.398 RSL: RESULT = {"Add1":"27.300"}
19:22:10.649 RSL: RESULT = {"Var2":"24.3"}
19:22:10.684 RUL: DS18B20-1#TEMPERATURE performs "backlog event t1=24.3"
19:22:10.699 RUL: DS18B20-2#TEMPERATURE performs "backlog event t2=24.7"
19:22:10.739 RSL: RESULT = {"Var1":"24.3"}
19:22:10.990 RSL: RESULT = {"Event":"Done"}
19:22:11.017 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:11.031 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:11.044 RSL: RESULT = {"Event":"Done"}
19:22:11.070 RUL: EVENT#T1 performs "Backlog Var1 24.3 ; ADD1 3"
19:22:11.088 RUL: EVENT#T1 performs "Backlog Var2 24.3 ; ADD2 1"
19:22:11.101 RSL: RESULT = {"Event":"Done"}
19:22:11.127 RUL: EVENT#T2>%VAR1% performs "POWER ON"
19:22:11.132 RSL: RESULT = {"POWER":"ON"}
19:22:11.135 RSL: POWER = ON
```

**Comment by sfromis** [comment]
While Tasmota is not multithreading, you should not treat rules as a programming language with a clear order of execution.

Anytime you do something which can satisfy another trigger, you should expect that trigger to fire before other things are executed. You need to be careful not to create race conditions. I see no signs of Tasmota having any race conditions, only your rules stepping on each other.

You also have the possibility of next iteration of the DS18B20 triggers firing before your convoluted rules have processed the prior one completely.

**Comment by steelman** [comment]
> You also have the possibility of next iteration of the DS18B20 triggers firing before your convoluted rules have processed the prior one completely.

That too crossed my mind, but the time constant of my bathroom isn't small enough to make subsequent readings differ significantly. While DS18B20 triggers were fired quicker than rules could be processed (after I disabled the ruleset there was quite a tail of log messages on the console indicating events still being processed), I still believe the race was between inequality checks and additions, because t2 being less than `t1 + 1` (thus triggering `POWER OFF` in one cycle) was greater than `t1` _before_ adding 3 which might triggered `POWER ON`.

 Is there a way to:

a) reduce sampling rate the sensors without using `tele-` to give time for the data processing rules to execute,
b) somehow mark chains of related rules that should be executed synchronously,
c) make `Backlog` in a single rule execute atomically so the comparisons of `event#t2` don't happen between assignments and additions (which I _believe_ took place)?

While `tele-` works quite OK (I mean, I don't get the flicker but it doesn't seem to fix the structure), both sensor reading events get triggered simultaneously and their `event#tX` rules still _may_ step on each other.

Besides, I'd like to react to temperature reaching thresholds as quick as possible, which d

PS. To make it clear, I didn't read the code yet and I may make wrong assumptions leading to wrong conclusions about stuff I observe.

**Reply by sfromis** [reply, marked as answer]
There is no concept of "related" rules, they are independent, except when they depend on the same trigger event (like the sensors cycle), or use features like one-shot or `break`. Hence no way of stipulating a flow.

You have quite a mess around processing of the `T1` trigger, the log very clearly shows more than one invocation of that before the Var commands come in later. IDK the background of that cookbook example, or how much it has even been tested, but it just does not look like good practice to me.

As a starting point, I'd just scrap the whole T1 thing, and replace by a single rule like:
`ON DS18B20-1#temperature DO Backlog0 Var1 %value%; ADD1 3; Var2 %value%; ADD2 1 ENDON`

Note that using `Backlog0` avoid the default `Backlog` inter-command delay of 200 ms.

**Reply by steelman** [reply]
Indeed, `Backlog0` alone fixes the mess even with a separate `event#t1` (which is redundant of course, I am updating the [pull request](https://github.com/tasmota/docs/pull/1358)).
